### PR TITLE
Feature/refuel service

### PR DESCRIPTION
**Pull request by Daan298261**
Up and running

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+.vs
```
Comment (markieo1) [inline .gitignore]: Gitignore incorrect see [VisualStudio.gitignore](https://github.com/github/gitignore/blob/master/VisualStudio.gitignore)

File: README.md [hunk 2]
```diff
@@ -1 +0,0 @@
-# RefuelService
```
Comment (markieo1) [inline README.md]: No need for this change

File: RefuelService.App/.gitignore [hunk 3]
```diff
@@ -0,0 +1,3 @@
+.vs
```
Comment (markieo1) [inline RefuelService.App/.gitignore]: File should not be needed

File: RefuelService.App/Messaging/RefuelEventHandler.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+
+
+        #region ---Outgoing events---
+        private async Task<bool> HandleServiceRequested(string message)
+        {
+            
+            Ship refueledShip = JsonSerializer.Deserialize<Ship>(message);
+            //lets say we executed the refueling instantly like a god we have no further stuff here IF it is the right service
+            if (refueledShip.ShipService.Name == "Refuel")
```
Comment (markieo1) [inline RefuelService.App/Messaging/RefuelEventHandler.cs]: We can use the guid for this. Every service will have it's own GUID. That way you can create an constant with a guid for your service and compare this to check if this service is requested. See the ContainerService for an example.

File: RefuelService.App/Properties/launchSettings.json [hunk 5]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (markieo1) [inline RefuelService.App/Properties/launchSettings.json]: Should not be in the repo. Fix by gitignore

File: RefuelService.Core/.gitignore [hunk 6]
```diff
@@ -0,0 +1,3 @@
+.vs
```
Comment (markieo1) [inline RefuelService.Core/.gitignore]: File not needed.

File: RefuelService.Core/Models/Ship.cs [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+{
+   public class Ship
+    {
+
+        [Key]
+        public string Serial { get; set; }
+        public string Name { get; set; }
+
+        [IgnoreDataMember]
+        public ShipService ShipService { get; set; }
```
Comment (markieo1) [inline RefuelService.Core/Models/Ship.cs]: Ship should not need to contain a ShipService.

File: RefuelService.Infrastructure/.gitignore [hunk 8]
```diff
@@ -0,0 +1,3 @@
+.vs
```
Comment (markieo1) [inline RefuelService.Infrastructure/.gitignore]: File not needed.

File: RefuelService.Infrastructure/Services/RefuelService.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+        public Task SendServiceCompletedAsync(Ship ship)
+        {
+            return Task.Run(async () =>
+            {
+              
+                ShipService shipService = new ShipService();
+                shipService.Name = "Refuel";
+                shipService.Id = new Guid();
+                shipService.Price = 2500;
```
Comment (markieo1) [inline RefuelService.Infrastructure/Services/RefuelService.cs]: This service should only do the Refueling. Not decided prices / ids / names and such. See ContainerService for example.

**Review by markieo1 — CHANGES_REQUESTED** [review]

File: RabbitMQTestEvents.txt [hunk 10]
```diff
@@ -0,0 +1,10 @@
+serviceRequested
```
Comment (Aplicticus) [inline RabbitMQTestEvents.txt]: Remove this file from Git.

File: RefuelService.App/Messaging/RefuelEventHandler.cs [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        {
+            switch (eventType)
+            {
+                case EventTypes.ShipDocked:
+                    {
+                        return await HandleShipDocked(message);
+                    }
+                case EventTypes.ShipUndocked:
+                    {
+
```
Comment (Aplicticus) [inline RefuelService.App/Messaging/RefuelEventHandler.cs:28]: Remove whitespace

File: RefuelService.App/Messaging/RefuelEventHandler.cs [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+                        return true;
+                    }
+            }
+
+            return true;
+        }
+
+        private async Task<bool> HandleShipUndocked(string message)
+        {
+
```
Comment (Aplicticus) [inline RefuelService.App/Messaging/RefuelEventHandler.cs:46]: Remove whitespaces

File: RefuelService.App/Messaging/RefuelEventHandler.cs [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+            Ship receivedShip = JsonSerializer.Deserialize<Ship>(message);
+            Ship existingShip = await _refuelService.GetShipAsync(receivedShip.Id);
+
+            await _refuelService.DeleteShipAsync(existingShip.Id);
+            //TODO? use enum to set ship as undocked so we dont have to delete and recreate it all the time?
+
+            Task.Run(() => _refuelService.SendShipUndockedAsync(existingShip));
```
Comment (Aplicticus) [inline RefuelService.App/Messaging/RefuelEventHandler.cs]: Why do you send event "UndockShip" in ShipUndocked? Refueling Service is not responsible for undocking ships.

File: RefuelService.App/Messaging/RefuelEventHandler.cs [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+            return true;
+        }
+
+        private async Task<bool> HandleShipDocked(string message)
+        {
+            //1. Deserialize ship
+            Ship receivedShip = JsonSerializer.Deserialize<Ship>(message);
```
Comment (Aplicticus) [inline RefuelService.App/Messaging/RefuelEventHandler.cs:57]: Use C# default conventions: 
For example: var receivedShip = JsonSerialier.Deserializer<Ship>(message);

File: RefuelService.App/Messaging/RefuelEventHandler.cs [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+            //2. Dump ship in db
+            Ship createdShip = await _refuelService.CreateShipAsync(receivedShip);
```
Comment (Aplicticus) [inline RefuelService.App/Messaging/RefuelEventHandler.cs:59]: Use C# default conventions:
For example: var receivedShip = JsonSerialier.Deserializer(message);

File: RefuelService.App/Messaging/RefuelEventHandler.cs [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+                    }
+            }
+
+            return true;
+        }
+
+        private async Task<bool> HandleShipUndocked(string message)
+        {
+
+            Ship receivedShip = JsonSerializer.Deserialize<Ship>(message);
```
Comment (Aplicticus) [inline RefuelService.App/Messaging/RefuelEventHandler.cs:47]: Use C# default conventions:
For example: var receivedShip = JsonSerialier.Deserializer(message);

File: RefuelService.App/Messaging/RefuelEventHandler.cs [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+            Ship existingShip = await _refuelService.GetShipAsync(receivedShip.Id);
```
Comment (Aplicticus) [inline RefuelService.App/Messaging/RefuelEventHandler.cs:48]: Use C# default conventions:
For example: var receivedShip = JsonSerialier.Deserializer(message);

File: RefuelService.App/Messaging/RefuelEventHandler.cs [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+        }
+
+        private async Task<bool> HandleShipDocked(string message)
+        {
+            //1. Deserialize ship
+            Ship receivedShip = JsonSerializer.Deserialize<Ship>(message);
+            //2. Dump ship in db
+            Ship createdShip = await _refuelService.CreateShipAsync(receivedShip);
+            //3. Send the docked event out
+            Task.Run(() => _refuelService.SendShipDockedAsync(createdShip));
```
Comment (Aplicticus) [inline RefuelService.App/Messaging/RefuelEventHandler.cs]: Why do you send event "DockShip" in ShipDocked? Refueling Service is not responsible for docking ships.

File: RefuelService.App/Messaging/RefuelEventHandler.cs [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+            //4. Do i really need to explain the below line?
+            return true;
+        }
+
+        private async Task<bool> HandleServiceRequested(string message)
+        {
+
```
Comment (Aplicticus) [inline RefuelService.App/Messaging/RefuelEventHandler.cs:65]: Remove whitespaces

File: RefuelService.App/Messaging/RefuelEventHandler.cs [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+            //musing servicerequest model now
+            var receivedShipService = JsonSerializer.Deserialize<ServiceRequest>(message);
+
+            //check if the service that is requested actually is refuelling.
+            if (receivedShipService.ServiceId == ShipServiceConstants.RefuelId)
+            {
+                Ship existingShip = await _refuelService.GetShipAsync(receivedShipService.ShipId);
```
Comment (Aplicticus) [inline RefuelService.App/Messaging/RefuelEventHandler.cs:72]: Use C# default conventions:
For example: var receivedShip = JsonSerialier.Deserializer(message);

File: RefuelService.Core/Messaging/IEventPublisher.cs [hunk 21]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace RefuelService.Core.Messaging
+{
+    public interface IEventPublisher
+    {
+        Task HandleEventAsync<T>(EventTypes eventType, T message);
```
Comment (Aplicticus) [inline RefuelService.Core/Messaging/IEventPublisher.cs:10]: Wrong function name. EventPublisher handles incoming events?

File: RefuelService.Core/Models/ServiceRequest.cs [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace RefuelService.Core.Models
+{
+    public class ServiceRequest
+    {
+        [Required]
+        public Guid ServiceId { get; set; }
+        [Required]
+        public Guid ShipId { get; set; }
+
```
Comment (Aplicticus) [inline RefuelService.Core/Models/ServiceRequest.cs:14]: Remove whitespaces

File: RefuelService.Core/Models/Ship.cs [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+        /// </summary>
+        [Required]
+        [Key]
+        public Guid Id { get; set; }
+
+        /// <summary>
+        /// Gets or sets the email.
+        /// </summary>
+        [Required]
+        public string Email { get; set; }
```
Comment (Aplicticus) [inline RefuelService.Core/Models/Ship.cs]: Remove email property

File: RefuelService.Core/Models/Ship.cs [hunk 24]
```diff
@@ -0,0 +1,27 @@
...
+
+        /// <summary>
+        /// Gets or sets the ship services.
+        /// </summary>
+        public List<ShipService> ShipServices { get; set; }
```
Comment (Aplicticus) [inline RefuelService.Core/Models/Ship.cs]: Remove this list of ShipServices

File: RefuelService.Core/Models/ShipServiceConstants.cs [hunk 25]
```diff
@@ -0,0 +1,16 @@
...
+
+namespace RefuelService.Core.Models
+{
+    public static class ShipServiceConstants
+    {
+        /// <summary>
+        /// Id of this service
+        /// </summary>
+        public static readonly Guid RefuelId = Guid.Parse("ff9f7c8c-92b6-4863-8451-de7c94389d53");
+
```
Comment (Aplicticus) [inline RefuelService.Core/Models/ShipServiceConstants.cs:13]: Remove whitespaces

File: RefuelService.Core/Repositories/IShipRepository.cs [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+
+        /// <summary>
+        /// Deletes the ship.
+        /// </summary>
+        /// <param name="Id">The identifier.</param>
+        /// <returns></returns>
+        Task DeleteShip(Guid Id);
+
+        /// <summary>
+		/// Gets the ship.
```
Comment (Aplicticus) [inline RefuelService.Core/Repositories/IShipRepository.cs:24]: Fix identation

File: RefuelService.Core/Services/IRefuelService.cs [hunk 27]
```diff
@@ -0,0 +1,59 @@
...
+        /// <param name="ship">The ship.</param>
+        /// <returns></returns>
+        Task<Ship> Refuel(Ship ship);
+
+        /// <summary>
+        /// Sends the ShipDocked Event
+        /// </summary>
+        /// <param name="ship">The ship that docked.</param>
+        /// <returns></returns>
+        Task<Ship> SendShipDockedAsync(Ship ship);
```
Comment (Aplicticus) [inline RefuelService.Core/Services/IRefuelService.cs]: Remove this function, RefuelService is not responsible for this action.

File: RefuelService.Core/Services/IRefuelService.cs [hunk 28]
```diff
@@ -0,0 +1,59 @@
...
+
+        /// <summary>
+        /// Sends the ShipUndocked Event
+        /// </summary>
+        /// <param name="ship">The ship that undocked.</param>
+        /// <returns></returns>
+        Task<Ship> SendShipUndockedAsync(Ship ship);
```
Comment (Aplicticus) [inline RefuelService.Core/Services/IRefuelService.cs]: Remove this function, RefuelService is not responsible for this action.

File: RefuelService.Infrastructure/Repositories/ShipRepository.cs [hunk 29]
```diff
@@ -0,0 +1,51 @@
...
+            await _database.SaveChangesAsync();
+
+            return newShip;
+        }
+
+        public Task<Ship> GetShip(Guid shipId)
+        {
+            return _database.Ships.LastOrDefaultAsync(r => r.Id == shipId);
+        }
+
```
Comment (Aplicticus) [inline RefuelService.Infrastructure/Repositories/ShipRepository.cs:47]: Remove white spaces / lines

File: RefuelService.Infrastructure/Services/RefuelService.cs [hunk 30]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        private readonly IShipRepository _shipRepository;
+        private readonly IEventPublisher _eventPublisher;
+
+        public RefuelService(IShipRepository shipRepository, IEventPublisher eventPublisher)
+        {
+            _shipRepository = shipRepository;
+            _eventPublisher = eventPublisher;
+        }
+
```
Comment (Aplicticus) [inline RefuelService.Infrastructure/Services/RefuelService.cs:20]: Remove white lines

File: RefuelService.Infrastructure/Services/RefuelService.cs [hunk 31]
```diff
@@ -0,0 +1,79 @@
...
+        public Task<Ship> GetShipAsync(Guid shipId)
+        {
+            return _shipRepository.GetShip(shipId);
+        }
+        public Task<Ship> Refuel(Ship ship)
+        {
+            //return shipservice
+            return Task.Run(async () =>
+            {
+
```
Comment (Aplicticus) [inline RefuelService.Infrastructure/Services/RefuelService.cs]: Remove white space / lines

File: RefuelService.Infrastructure/Services/RefuelService.cs [hunk 32]
```diff
@@ -0,0 +1,79 @@
...
+                //wait 2sec to spoof refuelling
+                await Task.Delay(2000);
+                //NOTE Do not send events out here, this is just the implementation of the refuelling!
+                return ship;
+            });
+        }
+
+        public Task<ServiceRequest> SendServiceCompletedAsync(ServiceRequest serviceRequest)
+        {
+            return Task.Run(async () =>
```
Comment (Aplicticus) [inline RefuelService.Infrastructure/Services/RefuelService.cs]: Remove Task.Run(async () => ). Just return the await. Set async before Task. Example: public async Task<x> name ().

File: RefuelService.Infrastructure/Services/RefuelService.cs [hunk 33]
```diff
@@ -0,0 +1,79 @@
...
+            {
+                    // do not change the serviceRequest object, if it enters this function it already has the right ship and service Id's
+                    await _eventPublisher.HandleEventAsync(EventTypes.ServiceCompleted, serviceRequest);
+                return serviceRequest;
+            });
+        }
+
+        public Task<Ship> SendShipDockedAsync(Ship ship)
+        {
+            return Task.Run(async () =>
```
Comment (Aplicticus) [inline RefuelService.Infrastructure/Services/RefuelService.cs]: Remove Task.Run(async () => ). Just return the await. Set async before Task. Example: public async Task<x> name ().

File: RefuelService.Infrastructure/Services/RefuelService.cs [hunk 34]
```diff
@@ -0,0 +1,79 @@
...
+            {
+                await _eventPublisher.HandleEventAsync(EventTypes.ShipDocked, ship);
+                return ship;
+            });
+        }
+
+        public Task<Ship> SendShipUndockedAsync(Ship ship)
+        {
+            return Task.Run(async () =>
```
Comment (Aplicticus) [inline RefuelService.Infrastructure/Services/RefuelService.cs]: Remove Task.Run(async () => ). Just return the await. Set async before Task. Example: public async Task<x> name ().

**Review by Aplicticus — CHANGES_REQUESTED** [review]

**Review by Aplicticus — APPROVED** [review]

File: RefuelService.Infrastructure/Database/RefuelDbContext.cs [hunk 35]
```diff
@@ -0,0 +1,17 @@
+﻿using Microsoft.EntityFrameworkCore;
+using RefuelService.Core.Models;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Refuel.Infrastructure.Database
```
Comment (markieo1) [inline RefuelService.Infrastructure/Database/RefuelDbContext.cs]: Namespace mismatch, should be RefuelService.Infrastructure.Database

**Review by markieo1 — CHANGES_REQUESTED** [review]

File: RefuelService.Infrastructure/Database/RefuelDbContext.cs (hunk 35 above)
Comment (Daan298261) [reply RefuelService.Infrastructure/Database/RefuelDbContext.cs]: fixed

**Review by markieo1 — APPROVED** [review]

**Pull request merged**